feat: Introduce HydratorClass on AttributeEntity

raffaelecarelle opened the pull request:
Attribute Entities do not have possibility to change the hydrator class.

raffaelecarelle committed:
Add support for defining and retrieving hydrator classes

This update introduces a new "hydrator_class" attribute and ensures it is properly defined in metadata and accessible via a getter method. This enhances flexibility by allowing custom hydrators to be specified for entities within the framework.

raffaelecarelle committed:
update docs

raffaelecarelle committed:
Add hydratorClass parameter to Entity.php docblock

Updated the docblock in Entity.php to include the hydratorClass parameter for better clarity. This improves type hinting and enhances code readability for developers.

h1k3r commented:
Hi @raffaelecarelle 

thanks a lot for your contribution! 💙 

As this is so sensitive part of the framework, would you mind adding few test cases covering your change?
You can find some examples in `tests/integration/Core/Framework/DataAbstractionLayer/AttributeEntityIntegrationTest.php`

If you would want some assistance, feel free to get back.

raffaelecarelle committed:
Add support for custom hydrator in integration tests

Introduced a new `DummyHydrator` and associated `AttributeEntityWithHydrator` entity for testing scenarios involving custom hydrators. Updated integration tests, configuration, and schema to include and validate the functionality. These changes ensure extensibility and improved test coverage for custom hydrator logic.

raffaelecarelle commented:
> Hi @raffaelecarelle
> 
> thanks a lot for your contribution! 💙
> 
> As this is so sensitive part of the framework, would you mind adding few test cases covering your change? You can find some examples in `tests/integration/Core/Framework/DataAbstractionLayer/AttributeEntityIntegrationTest.php`
> 
> If you would want some assistance, feel free to get back.

Hello @h1k3r, you're right! I wasn't sure where and how...but now is clear, thank you!

raffaelecarelle commented:
@h1k3r I've introduce tests for custom hydrator class

mitelg commented on tests/integration/Core/Framework/DataAbstractionLayer/AttributeEntityIntegrationTest.php:
```diff
@@ -789,6 +800,29 @@ public function testManyToManyVersioned(): void
+        $search = $this->repository('attribute_entity_with_hydrator')
+            ->search(new Criteria([$ids->get('first-key')]), Context::createDefaultContext());
+
+        /** @var AttributeEntityWithHydrator $record */
```
please do not use var annotations. Especially in tests you can simply assert the type. also see https://github.com/shopware/shopware/blob/trunk/coding-guidelines/core/writing-code-for-static-analysis.md

raffaelecarelle replied:
sorry for that! I've fixed

raffaelecarelle committed:
Add assertion for record type in integration test

Ensure the retrieved record is of the expected type using `assertInstanceOf`. This improves test reliability and prevents potential type-related issues.

mitelg commented on changelog/_unreleased/2025-01-08-introduce-hydrator-class-attribute-entity.md:
```diff
@@ -0,0 +1,11 @@
+---
+title: Introduce Hydrator class on Attribute Entity 
+issue: NEXT-00000
```
this is not needed anymore

raffaelecarelle replied:
ok, so can I remove the file?

mitelg replied:
no, just the line about the issue. sorry, that was not clear enough

raffaelecarelle replied:
no problem, I removed issue line

mitelg commented on changelog/_unreleased/2025-01-08-introduce-hydrator-class-attribute-entity.md:
```diff
@@ -0,0 +1,11 @@
+---
+
+# Core
+ * Changed `\Shopware\Core\Framework\DataAbstractionLayer\Attribute\Entity` adding `$hydratoClass` parameter
```
```suggestion
 * Changed `\Shopware\Core\Framework\DataAbstractionLayer\Attribute\Entity` to allow defining a hydrator class
```

mitelg commented on changelog/_unreleased/2025-01-08-introduce-hydrator-class-attribute-entity.md:
```diff
@@ -0,0 +1,11 @@
+
+# Core
+ * Changed `\Shopware\Core\Framework\DataAbstractionLayer\Attribute\Entity` adding `$hydratoClass` parameter
+ * Changed `\Shopware\Core\Framework\DataAbstractionLayer\AttributeEntityCompiler` which process `$hydratorClass` parameter
```
we do not need to document implementation details in the changelog

mitelg commented on changelog/_unreleased/2025-01-08-introduce-hydrator-class-attribute-entity.md:
```diff
@@ -0,0 +1,11 @@
+# Core
+ * Changed `\Shopware\Core\Framework\DataAbstractionLayer\Attribute\Entity` adding `$hydratoClass` parameter
+ * Changed `\Shopware\Core\Framework\DataAbstractionLayer\AttributeEntityCompiler` which process `$hydratorClass` parameter
+ * Changed `\Shopware\Core\Framework\DataAbstractionLayer\AttributeEntityDefinition` overriding `getHydratorClass` that returns the hydrator class passed on Attribute 
```
same as above

raffaelecarelle committed:
Refine changelog entry for hydrator class changes.

Updated the changelog to clarify changes related to the hydrator class in `Attribute\Entity`. Consolidated redundant entries to improve readability and precision.

raffaelecarelle committed:
Remove issue placeholder from changelog entry.

The issue field was removed as it was left as a placeholder. This change ensures the changelog entry is cleaner and more accurate.

h1k3r commented on tests/integration/Core/Framework/DataAbstractionLayer/fixture/AttributeEntityWithHydrator.php:
```diff
@@ -0,0 +1,23 @@
+/**
+ * @internal
+ */
+#[Entity('attribute_entity_with_hydrator', hydratorClass: DummyHydrator::class)]
```
Currently few tests are failing: https://github.com/shopware/shopware/actions/runs/12765559819/job/35580098527?pr=6068

To fix `EntityDefinitionHasSinceTest` you may add since parameter:
```suggestion
#[Entity('attribute_entity_with_hydrator', since: '6.6.10.0', hydratorClass: DummyHydrator::class)]
```
To fix `ApiAwareTest` you may change `tests/integration/Core/Framework/Api/fixtures/api-aware-fields.json` (use attribute_entity.* fields as example)

raffaelecarelle replied:
Hi @h1k3r, I fixed now. Thank you

raffaelecarelle committed:
Add versioning to AttributeEntityWithHydrator entity

Included a new "since" metadata attribute to specify the version in which the entity was added. Updated fixture files to include additional fields for the entity in integration tests.

h1k3r commented on src/Core/Framework/DataAbstractionLayer/Attribute/Entity.php:
```diff
@@ -24,6 +26,7 @@ public function __construct(
         public ?string $parent = null,
         public ?string $since = null,
         public string $collectionClass = EntityCollection::class,
+        public string $hydratorClass = EntityHydrator::class,
```
Strictly speaking this can be a break if the class is extended somewhere.

But the entity attribute was never intended to be extended, and probability that it was done for some reason is low. After short discussion in the team we decided that this would be acceptable change. Would you be so kind to mark this class as final and add relevant configuration to .bc-exclude.php to make bc checker pass (https://github.com/shopware/shopware/actions/runs/12786506757/job/35643703712?pr=6068)

If you have other questions about making pipeline green - feel free to reach out!

raffaelecarelle replied:
Ok I've done as you suggest. 

I don´t think the [PHP checks / PHPUnit for unit (pull_request)](https://github.com/shopware/shopware/actions/runs/12787266449/job/35646091929?pr=6068) that is failed is related to this PR. Can you help me?

h1k3r replied:
Thank you!
I've updated the branch with the latest changes in trunk - those test should pass now.

raffaelecarelle committed:
Mark Entity class as final and update BC exclude list.

The Entity class was marked as final to prevent further extension and ensure immutability. Additionally, the BC exclude list was updated to reflect this change and avoid backward compatibility issues.

raffaelecarelle committed:
Remove unused field from API-aware fields fixture

The "attribute_entity_with_hydrator.number" field was removed from `api-aware-fields.json` as it is no longer in use. This reduces clutter and keeps the fixture file clean and relevant.

h1k3r commented on .bc-exclude.php:
```diff
@@ -13,6 +13,7 @@
         '**src/Core/Framework/Script/ServiceStubs.php', // never intended to be extended
         '**/src/Core/Framework/App/Source/AbstractTemporaryDirectoryFactory.php', // dropped (not released yet)
         '**/src/Core/Framework/App/Source/TemporaryDirectoryFactory.php', // dropped decorator (not released yet)
+        '**/src/Core/Framework/DataAbstractionLayer/Attribute/Entity.php', // never intended to be extended
```
I would not exclude the whole file, but rather add to the errors section next lines:
```
// Was not intended to be extended, declared as final
'Class Shopware\\\\Core\\\\Framework\\\\DataAbstractionLayer\\\\Attribute\\\\Entity became final',
'Parameter hydratorClass was added to Method __construct\(\) of class Shopware\\\\Core\\\\Framework\\\\DataAbstractionLayer\\\\Attribute\\\\Entity',
```

You can test it using `composer run bc-check -- --from="origin/trunk"`

raffaelecarelle replied:
alright! I've done :)

raffaelecarelle committed:
Update .bc-exclude.php to reflect Entity class changes

Removed Entity.php from excludes as it is now final. Added corresponding change notes for the Entity class final declaration and constructor update. This ensures backward compatibility rules are maintained.

h1k3r committed:
Merge branch 'trunk' into entity-attribute-hydrator-class

mitelg commented:
thanks for your contribution @raffaelecarelle 👍🎉💙
